store/v2/root/store_test.go

```diff
@@ -0,0 +1,958 @@
+	require.Panics(t, func() { store.Get([]byte{}) }, "Get(empty key) should panic")
+	require.Panics(t, func() { store.Has(nil) }, "Has(nil key) should panic")
+	require.Panics(t, func() { store.Has([]byte{}) }, "Has(empty key) should panic")
+	require.Panics(t, func() { store.Set(nil, []byte("value")) }, "Set(nil key) should panic")
+	require.Panics(t, func() { store.Set([]byte{}, []byte("value")) }, "Set(empty key) should panic")
+	require.Panics(t, func() { store.Set([]byte("key"), nil) }, "Set(nil value) should panic")
+	sub := store.(*substore)
+	sub.indexBucket = rwCrudFails{sub.indexBucket, nil}
+	require.Panics(t, func() {
+		store.Set([]byte("key"), []byte("value"))
+	}, "Set() when index fails should panic")
+}
+
+func TestConstructors(t *testing.T) {
+	db := memdb.NewDB()
+
+	store, err := NewStore(db, simpleStoreConfig(t))
+	require.NoError(t, err)
+	_ = store.GetKVStore(skey_1)
+	store.Commit()
+	require.NoError(t, store.Close())
+
+	// Loading with an initial version beyond the lowest should error
+	opts := StoreConfig{InitialVersion: 5, Pruning: types.PruneNothing}
+	store, err = NewStore(db, opts)
+	require.Error(t, err)
+	db.Close()
+
+	store, err = NewStore(dbVersionsFails{memdb.NewDB()}, DefaultStoreConfig())
+	require.Error(t, err)
+	store, err = NewStore(db, StoreConfig{StateCommitmentDB: dbVersionsFails{memdb.NewDB()}})
+	require.Error(t, err)
+
+	// can't use a DB with open writers
+	db = memdb.NewDB()
+	merkledb := memdb.NewDB()
+	w := db.Writer()
+	store, err = NewStore(db, DefaultStoreConfig())
+	require.Error(t, err)
+	w.Discard()
+	w = merkledb.Writer()
+	store, err = NewStore(db, StoreConfig{StateCommitmentDB: merkledb})
+	require.Error(t, err)
+	w.Discard()
+
+	// can't use DBs with different version history
+	merkledb.SaveNextVersion()
+	store, err = NewStore(db, StoreConfig{StateCommitmentDB: merkledb})
+	require.Error(t, err)
+	merkledb.Close()
+
+	// can't load existing store when we can't access the latest Merkle root hash
+	store, err = NewStore(db, simpleStoreConfig(t))
+	require.NoError(t, err)
+	store.Commit()
+	require.NoError(t, store.Close())
+	// ...whether because root is misssing
+	w = db.Writer()
+	s1RootKey := append(contentPrefix, substorePrefix(skey_1.Name())...)
+	s1RootKey = append(s1RootKey, merkleRootKey...)
+	w.Delete(s1RootKey)
+	w.Commit()
+	db.SaveNextVersion()
+	store, err = NewStore(db, DefaultStoreConfig())
+	require.Error(t, err)
+	// ...or, because of an error
+	store, err = NewStore(dbRWCrudFails{db}, DefaultStoreConfig())
+	require.Error(t, err)
+}
+
+func TestIterators(t *testing.T) {
+	_, store := newSubStoreWithData(t, memdb.NewDB(), map[string]string{
+		string([]byte{0x00}):       "0",
+		string([]byte{0x00, 0x00}): "0 0",
+		string([]byte{0x00, 0x01}): "0 1",
+		string([]byte{0x00, 0x02}): "0 2",
+		string([]byte{0x01}):       "1",
+	})
+
+	var testCase = func(t *testing.T, iter types.Iterator, expected []string) {
```

Comment:
Can this take a description?

Reply:
I could add it, but I'm not sure what a useful description would be here